MericCelik/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,291 @@
+     * Best O(1)(When target is the root)--- N is the node count of the tree.
+     * @param target : node whose ancestors are found.
+     * @return arraylist of ancestors.
+     * */
+
+    public ArrayList<BinaryTreeNode<T>> findAncestors(BinaryTreeNode<T> target){
+        ArrayList<BinaryTreeNode<T>> ancestors = new ArrayList<>();
+        BinaryTreeNode<T> temp = target;
+        while(temp.getParent()!=null){
+            ancestors.add(temp.getParent());
+            temp = temp.getParent();
+        }
+        return ancestors;
+    }
+
+    /**Calls findAncestors method and print the items of returned arraylist.
+     * It checks if the input is valid, tree is not empty, target is not the root and
+     * the target is in the tree. If these conditions are not hold, it prints suitable
+     * warning message to the standart output.
+     * Complexities: O(1) when warnings occurs, otherwise depend on findsAncestors method.
+     * @param target : node whose ancestors are found.
+     * @return nothing.
+     * */
+
+    public void printAncestors(BinaryTreeNode<T> target){
+        if(target==null){
+            System.out.println("There is no target node!");
+            return;
+        }
+        if(this.getRoot()==null){
+            System.out.println("The tree is empty!");
+            return;
+        }
+
+        if(target.equals(this.root)){
+            System.out.println("Target node is the root of the tree, it has no parent!");
+            return;
+        }
+
+        ArrayList<BinaryTreeNode<T>> ancestors = findAncestors(target);
+        int ancestor_count = ancestors.size();
+        if(ancestor_count==0 || !this.getRoot().equals(ancestors.get(ancestor_count-1))){
+            System.out.println("This node is not in the binary tree!");
+            return;
+        }
+
+        for(int i=0; i<ancestor_count;i++)
+            System.out.print(ancestors.get(i).getKey() + " ");
+        System.out.println();
+    }
+
+    /**Calls the findAncestors method for two nodes and finds their lowest common ancestor by comparing
+     * resulting arrays. If one of them is in other's resulting array, this mean it is the ancestor of the other.
+     * Else if the array sizes are equal, it means they are in the same level. It traverse arrays simultaneously
+     * until it finds first common item. This is the lowest common ancestor.
+     * Else if the array sizes are not equal, it discards the deeper node's extra ancestors. It does it by incrementing
+     * the index value for the array traverse.
+     * The order of the nodes are not important.
+     * Complexities: O(1) when warnings occurs, otherwise depend on findsAncestors method.
+     * @param node1 : first node to consider
+     * @param node2 : other node to consider
+     * @return lowest common ancestor node
+     * */
+
+    public BinaryTreeNode<T> lowestCommonAncestor(BinaryTreeNode<T> node1, BinaryTreeNode<T> node2){
```

Comment:
In these methods you already take the Nodes as input. Could you write a wrapper function for these that take keys instead (T) and first find these nodes in the tree?

Reply:
I added find() and findNodeWithDFS() functions. I had missed that point, thank you! :smile: